### ISSUE TITLE
java.lang.UnsatisfiedLinkError: nativeOpen on 1.2 and Android SDK15

### ISSUE THREAD

**Issue by serge-coffeeholic** [couchbase/couchbase-lite-android#792]
After upgrade to 1.2 observe the following exception on Android SDK 15 (emulator) with this code:
`new Manager(new AndroidContext(context), Manager.DEFAULT_OPTIONS);`

```
java.lang.UnsatisfiedLinkError: nativeOpen
    at com.couchbase.lite.internal.database.sqlite.SQLiteConnection.nativeOpen(Native Method)
    at com.couchbase.lite.internal.database.sqlite.SQLiteConnection.open(SQLiteConnection.java:194)
    at com.couchbase.lite.internal.database.sqlite.SQLiteConnection.open(SQLiteConnection.java:178)
    at com.couchbase.lite.internal.database.sqlite.SQLiteConnectionPool.openConnectionLocked(SQLiteConnectionPool.java:461)
    at com.couchbase.lite.internal.database.sqlite.SQLiteConnectionPool.open(SQLiteConnectionPool.java:191)
    at com.couchbase.lite.internal.database.sqlite.SQLiteConnectionPool.open(SQLiteConnectionPool.java:183)
    at com.couchbase.lite.internal.database.sqlite.SQLiteDatabase.openInner(SQLiteDatabase.java:814)
    at com.couchbase.lite.internal.database.sqlite.SQLiteDatabase.open(SQLiteDatabase.java:798)
    at com.couchbase.lite.internal.database.sqlite.SQLiteDatabase.openDatabase(SQLiteDatabase.java:693)
    at com.couchbase.lite.storage.SQLiteStorageEngineBase.open(SQLiteStorageEngineBase.java:62)
    at com.couchbase.lite.DatabaseUpgrade.importData(DatabaseUpgrade.java:81)
    at com.couchbase.lite.Manager.upgradeDatabase(Manager.java:830)
    at com.couchbase.lite.Manager.upgradeOldDatabaseFiles(Manager.java:808)
    at com.couchbase.lite.Manager.<init>(Manager.java:130)
```

**Comment by hideki** [comment]
Logcat
```
02-25 09:53:35.880 646-661/? W/Database: mustExist is true and db (cblite-test) does not exist
02-25 09:53:35.880 646-661/? D/dalvikvm: Trying to load lib /data/data/com.couchbase.lite.test/lib/libCouchbaseLiteJavaForestDB.so 0x4160eea0
02-25 09:53:35.920 646-661/? D/dalvikvm: Added shared lib /data/data/com.couchbase.lite.test/lib/libCouchbaseLiteJavaForestDB.so 0x4160eea0
02-25 09:53:35.941 646-661/? D/dalvikvm: No JNI_OnLoad found in /system/lib/libsqlite.so 0x4160eea0, skipping init
02-25 09:53:35.941 646-661/? D/dalvikvm: Trying to load lib /data/data/com.couchbase.lite.test/lib/libcbljavasqlitedefault.so 0x4160eea0
02-25 09:53:35.941 646-661/? E/Database: Cannot load native library
02-25 09:53:35.950 646-661/? D/dalvikvm: Trying to load lib /data/data/com.couchbase.lite.test/lib/libcbljavasqlitedefault.so 0x4160eea0
02-25 09:53:35.950 646-661/? E/Database: Cannot load native library
02-25 09:53:35.950 646-661/? W/dalvikvm: No implementation found for native Lcom/couchbase/lite/internal/database/sqlite/SQLiteConnection;.nativeOpen (Ljava/lang/String;ILjava/lang/String;ZZ)J
02-25 09:53:35.950 646-661/? I/TestRunner: failed: test12_AllDocumentsBySequence(com.couchbase.lite.ApiTest)
02-25 09:53:35.960 646-661/? I/TestRunner: ----- begin exception -----
02-25 09:53:36.001 646-661/? I/TestRunner: ----- end exception -----
02-25 09:53:36.030 646-649/? D/dalvikvm: GC_CONCURRENT freed 387K, 4% free 12933K/13383K, paused 5ms+2ms
02-25 09:53:36.030 646-661/? I/TestRunner: finished: test12_AllDocumentsBySequence(com.couchbase.lite.ApiTest)
```

StackTrace
```
Testing started at 9:53 AM ...
Target device: Nexus_4_API_15_arm [emulator-5554]
No apk changes detected since last installation, skipping installation of /Users/hideki/github/couchbase-lite-android/build/outputs/apk/couchbase-lite-android-debug-androidTest-unaligned.apk
Force stopping package: com.couchbase.lite.test
DEVICE SHELL COMMAND: am force-stop com.couchbase.lite.test
Running tests
Test running started
java.lang.UnsatisfiedLinkError: nativeOpen
at com.couchbase.lite.internal.database.sqlite.SQLiteConnection.nativeOpen(Native Method)
at com.couchbase.lite.internal.database.sqlite.SQLiteConnection.open(SQLiteConnection.java:194)
at com.couchbase.lite.internal.database.sqlite.SQLiteConnection.open(SQLiteConnection.java:178)
at com.couchbase.lite.internal.database.sqlite.SQLiteConnectionPool.openConnectionLocked(SQLiteConnectionPool.java:461)
at com.couchbase.lite.internal.database.sqlite.SQLiteConnectionPool.open(SQLiteConnectionPool.java:191)
at com.couchbase.lite.internal.database.sqlite.SQLiteConnectionPool.open(SQLiteConnectionPool.java:183)
at com.couchbase.lite.internal.database.sqlite.SQLiteDatabase.openInner(SQLiteDatabase.java:814)
at com.couchbase.lite.internal.database.sqlite.SQLiteDatabase.open(SQLiteDatabase.java:798)
at com.couchbase.lite.internal.database.sqlite.SQLiteDatabase.openDatabase(SQLiteDatabase.java:693)
at com.couchbase.lite.storage.SQLiteStorageEngineBase.open(SQLiteStorageEngineBase.java:62)
at com.couchbase.lite.store.SQLiteStore.open(SQLiteStore.java:183)
at com.couchbase.lite.Database.open(Database.java:1150)
at com.couchbase.lite.Manager.openDatabase(Manager.java:336)
at com.couchbase.lite.Manager.getDatabase(Manager.java:300)
at com.couchbase.lite.LiteTestCaseWithDB.ensureEmptyDatabase(LiteTestCaseWithDB.java:245)
at com.couchbase.lite.LiteTestCaseWithDB.startDatabase(LiteTestCaseWithDB.java:230)
at com.couchbase.lite.LiteTestCaseWithDB.setUp(LiteTestCaseWithDB.java:132)
at com.couchbase.lite.LiteTestCase.runBare(LiteTestCase.java:20)
at com.couchbase.lite.LiteTestCaseWithDB.runBare(LiteTestCaseWithDB.java:88)
at android.test.AndroidTestRunner.runTest(AndroidTestRunner.java:169)
at android.test.AndroidTestRunner.runTest(AndroidTestRunner.java:154)
at android.test.InstrumentationTestRunner.onStart(InstrumentationTestRunner.java:545)
at android.app.Instrumentation$InstrumentationThread.run(Instrumentation.java:1551)

Finish
```

**Comment by pasin** [comment]
> libsqlite.so exists, is it does not have JNI_OnLoad() method?
libsqlite.so is sqlite not JNI.

**Comment by pasin** [comment]
I guessed on API 15, it couldn't load system sqlite.

**Comment by hideki** [comment]
https://github.com/couchbase/couchbase-lite-android/issues/434

As described in linked ticket, I guess Android 4.0.3 might try to load armeabi JNI library instead of armeabi-v7a. 

We don't provide armeabi JNI for forestdb, it might explain why loading forestdb JNI, but not SQLite Jni.

**Comment by filipesperandio** [comment]
Is there a workaround?

**Comment by zgramana** [comment]
@filipesperandio: The workaround should be using an emulator that is 4.1+ / API 16+, as the bug is in Android 4.0.3 itself.

We implemented a workaround previously, but when we overhauled our storage code we didnt also re-implement the fix. This will get fixed in an upcoming patch release.

**Comment by filipesperandio** [comment]
I am targeting devices running android 4.0.4 and got this very issue. If is there anything I can patch around to stick with couchbase I would love it, otherwise I think will have to reimplement my persistence layer with SQLite itself. =\

**Comment by zgramana** [comment]
@filipesperandio try using [the APK split technique](http://tools.android.com/tech-docs/new-build-system/user-guide/apk-splits). This should bypass the Android bug by only including native images for one platform per APK.

**Comment by filipesperandio** [comment]
No much luck so far.
Tried this:
```
splits {
        abi {
            enable true
            reset()
            include 'armeabi-v7a', 'x86'
            universalApk false
        }
    }
```
But error still happens on that device. Should it be something slightly different from that? Not sure if I got the problem right. Is it an issue the way android 4.0.4 loads `abi`s native code? (sort of)

**Comment by hideki** [comment]
Couchbase Lite Android JNI loads system sqlite3.so from Android, and call `sqlite3_db_readonly()` method which was added from 3.7.11. However, Android API 15 or earlier uses SQLite 3.7.4 or earlier. `System.loadLibrary(<CBL JNI library name>)` can identify unsatisfied method. Then throws Exception.

We Fixed this by commented out the lines to check if SQLite database is readonly.
PR: https://github.com/couchbase/couchbase-lite-java-native/pull/40

Note: I tried some solutions by checking if function pointer is null or checking SQLite version to identify `sqlite3_db_readonly()` exists. However, `System.loadLibrary()` throws UnsatisfiedLinkError if code calls this method.

**Comment by hideki** [comment]
Following is build from master branch. This includes the fix.
http://files.couchbase.com/maven2/com/couchbase/lite/couchbase-lite-android/0.0.0-605/

**Comment by filipesperandio** [comment]
Cool. Thanks!

**Comment by dougkoellmer** [comment]
Hello all. My first time posting an issue to this repository. Thanks for all your work so far on this library.

I'm still getting this crash on a Kindle Fire. I've attached the output of `adb shell getprop` in case it's helpful.

[kindle_props.txt](https://github.com/couchbase/couchbase-lite-android/files/178658/kindle_props.txt)

I've tried several builds, including the latest from master, the 605 referenced above, and 1.2.1 (was doing final release testing with 1.2.0 when this came up).

The stack trace looks the same as the ones pasted above, but just in case I've attached mine as well.

[stack_trace.txt](https://github.com/couchbase/couchbase-lite-android/files/178668/stack_trace.txt)

Thanks for any help!
